### PR TITLE
Adding Euclidean Alignment implementation

### PR DESCRIPTION
**Pull request by brunaafl**
Hello!

I am adding an implementation of the Euclidean Alignment (He H. et al - 2019) used on Junqueira, B et al. (2024). The methods depend directly on the covariance estimators present on Pyriemann, and Sylvain thought about what would be a nice addition for Pyriemann.

In this re-implementation, I defined it as a scikit-learn transformation, and in a way that is not necessary to define a ‘target’ domain before. 

It would be easier to add this transformation as a preprocessing step in a pipeline, and also to integrate with MOABB, being able to use the pre-defined evaluation objects. 

This class is also adapted to align both raw trials and covariance matrices and to define the type of covariance estimator, in the case the input is raw data. Therefore, it can be used with different classifiers.

Since we were working with this transformation, we thought adding an implementation and a tutorial here in PyRiemann would be nice, but I would like to check first if the tutorial would be welcome.

### CODE REVIEW

**Comment by brunaafl** [comment]
@qbarthelemy
Fyi: @sylvchev, @bruAristimunha

I'm also preparing some test scrips to add for testing the transformation.

File: pyriemann/transfer/_euclidean.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+from scipy.linalg import sqrtm, inv
+from numpy import iscomplexobj, real, any, isfinite
+from sklearn.base import BaseEstimator, TransformerMixin
+from pyriemann.estimation import covariances
+
+from pyriemann.utils.mean import mean_covariance
+from pyriemann.transfer import decode_domains
+
+
+def compute_ref_euclidean(data):
```
Comment (gcattan) [inline pyriemann/transfer/_euclidean.py:12]: I think this method should be private.

File: pyriemann/transfer/_euclidean.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    data: numpy array.
+    It can be either raw trials with shape (n, c, t) or covariances (n, c, c)
+    Reference trials to compute the alignment matrix
+
+    Returns
+    -------
+    r_ea: numpy array with shape (c, c)
+    Reference matrix
+
+    '''
```
Comment (gcattan) [inline pyriemann/transfer/_euclidean.py:27]: You can remove this, if your method is private.

File: pyriemann/transfer/_euclidean.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+    mean = mean_covariance(data, metric='euclid')
+
+    compare = np.allclose(mean, np.identity(mean.shape[0]))
+
+    if not compare:
+        if iscomplexobj(mean):
+            print("covariance matrix problem")
```
Comment (gcattan) [inline pyriemann/transfer/_euclidean.py:35]: Here and below, I suggest to use `warns` instead of `print`.

File: pyriemann/transfer/_euclidean.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+    This transformation accept both raw signals and covariance matrices as
+    data inputs. For raw trials, you can determine the estimator using the
+    'estimtor' parameter.
+
+    .. note::
+       Using .fit() and then .transform() will give different results than
+       .fit_transform(). In fact, .fit_transform() should be applied on the
+       training dataset (target and source) and .transform() on the test
+       partition of the target dataset.
```
Comment (gcattan) [inline pyriemann/transfer/_euclidean.py:71]: I think this is the default behavior for transformer_mixin.
https://github.com/scikit-learn/scikit-learn/blob/f07e0138bfee41cd2c0a5d0251dc3fe03e6e1084/sklearn/base.py#L1043

File: pyriemann/transfer/_euclidean.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+    Parameters
+    ----------
+
+    dtype: str
+    Type of input data, it can be 'covmat' for covariance or 'raw' for raw trials
```
Comment (gcattan) [inline pyriemann/transfer/_euclidean.py:77]: I think it should be an indentation here.

File: pyriemann/transfer/_euclidean.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+    estimator: str
+    If dtype=='raw', it represents the covariance estimator
```
Comment (gcattan) [inline pyriemann/transfer/_euclidean.py:80]: And if it is "covmat" or something else?

File: pyriemann/transfer/_euclidean.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        Dictionary containing the covariance matrices for each domain.
+        Order of the list is the same as 'domains'
+
+    References
+    ----------
+    .. [1] `Transfer Learning for Brain-Computer Interfaces:
+        A Euclidean Space Data Alignment Approach
+        <https://arxiv.org/abs/1808.05464>`_
+        He He and Dongrui Wu, IEEE Transactions on Biomedical Engineering, 2019
+
```
Comment (gcattan) [inline pyriemann/transfer/_euclidean.py:98]: Can you add:
```
Notes
-------
..versionadded:: 0.6.0
```

File: pyriemann/transfer/_euclidean.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        return X_align
+
+    def _transform_cov(self, X, d):
+        """ Compute covariance matrices of trials in array X and save"""
+
+        cov = covariances(X, estimator=self.estimator)
+        self._cov[d] = cov
+        return cov
+
+    def fit(self, X, y_enc):
```
Comment (gcattan) [inline pyriemann/transfer/_euclidean.py:150]: `fit` method is public, it should be documented.
Something along these lines (except you accept both epochs and covariance matrices):

https://github.com/pyRiemann/pyRiemann/blob/c44b924573dc2f4b4a9efb328d18636b8baa31d5/pyriemann/tangentspace.py#L83

File: pyriemann/transfer/_euclidean.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+        _, _, domains = decode_domains(X, y_enc)
+
+        self._fit_sw(X, domains)
+
+        return self
+
+    def transform(self, X, y_enc=None):
+        """
+        Used for transforming the target data after fitting the calibration
```
Comment (gcattan) [inline pyriemann/transfer/_euclidean.py:162]: Should be also documented, as it is a public method.

**Review by gcattan — CHANGES_REQUESTED** [review]
Hi @brunaafl !
Nice work :) IMO, I think it could be a great idea to add a tutorial too. 

Below, some minor comments on the form.

File: pyriemann/transfer/_euclidean.py (hunk 7 above)
Comment (gcattan) [reply pyriemann/transfer/_euclidean.py:98]: Can you also add the reference to your paper?

File: pyriemann/transfer/_euclidean.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+
+def compute_ref_euclidean(data):
+    '''
+    Calculate the reference matrix for computing the Euclidean alignment,
+
+    Parameters
+    ----------
+    data: numpy array.
+    It can be either raw trials with shape (n, c, t) or covariances (n, c, c)
```
Comment (VascoSch92) [inline pyriemann/transfer/_euclidean.py:19]:
```suggestion
    It can be either raw trials with shape (n, c, t) or covariances with shape (n, c, c)
```

File: pyriemann/transfer/_euclidean.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+            print("covariance matrix problem")
+        if iscomplexobj(sqrtm(mean)):
+            print("covariance matrix problem sqrt")
+
+        r_ea = inv(sqrtm(mean))
+
+        if iscomplexobj(r_ea):
+            print("WARNING! Covariance matrix was not SPD somehow. " +
+                  "Can be caused by running ICA-EOG rejection, if " +
+                  "not, check data!!")
```
Comment (VascoSch92) [inline pyriemann/transfer/_euclidean.py:44]:
```suggestion
            warns(
                "Covariance matrix was not SPD somehow. \n"
                 "Can be caused by running ICA-EOG rejection, if " 
                 "not, check the data!"
             )
```

File: pyriemann/transfer/_euclidean.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+    .. [1] `Transfer Learning for Brain-Computer Interfaces:
+        A Euclidean Space Data Alignment Approach
+        <https://arxiv.org/abs/1808.05464>`_
+        He He and Dongrui Wu, IEEE Transactions on Biomedical Engineering, 2019
+
+    """
+    def __init__(self, estimator='lwf', dtype='covmat'):
+        self.estimator = estimator
+        self.dtype = dtype
+        self._ref = list()
```
Comment (VascoSch92) [inline pyriemann/transfer/_euclidean.py:103]:
```suggestion
        self._ref = []
```
just to be coherent with `self._cov` ;-)

File: pyriemann/transfer/_euclidean.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+        self._cov = {}
+
+    def _compute_ref(self, X):
+        ref = compute_ref_euclidean(X)
+
+        self._ref.append(ref)
+
+    def _fit_sw(self, X, domains):
+
+        for i in range(len(np.unique(domains))):
```
Comment (VascoSch92) [inline pyriemann/transfer/_euclidean.py:113]: I think here you can just loop on` np.unique(domains)`, i.e.,
```python
for d in np.unique(domains):
   ...
```

File: pyriemann/transfer/_euclidean.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+            if self.dtype != 'covmat':
+                X_d = self._transform_cov(X_d, d)
+
+            self._compute_ref(X_d)
+
+    def _transform_sw(self, X, domains):
+        """Align each domain"""
+
+        X_align = []
+        for i in range(len(np.unique(domains))):
```
Comment (VascoSch92) [inline pyriemann/transfer/_euclidean.py:128]: same as above

File: pyriemann/transfer/_euclidean.py (hunk 4 above)
Comment (brunaafl) [reply pyriemann/transfer/_euclidean.py]: You might be right. I was not sure if it was the case, so I put this comment. Since I was following [https://github.com/pyRiemann/pyRiemann/blob/master/pyriemann/transfer/_estimators.py#L89](https://github.com/pyRiemann/pyRiemann/blob/c44b924573dc2f4b4a9efb328d18636b8baa31d5/pyriemann/transfer/_estimators.py#L89) as a reference to my code, I put the same note.

File: pyriemann/transfer/_euclidean.py [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+        Used for fitting and transforming 'train' data.
+        """
+
+        self.fit(X, y_enc)
+
+        _, _, domains = decode_domains(X, y_enc)
+
+        X_align = self._transform_sw(X, domains)
+
+        return X_align
```
[the remark this answers is not in the document]
Comment (brunaafl) [reply pyriemann/transfer/_euclidean.py]: I did this so I can could separate `fit_transform` as the method to be applied during the training/fitting, while `transform` during inference in the target set.

File: pyriemann/transfer/_euclidean.py (hunk 13 above)
Comment (brunaafl) [reply pyriemann/transfer/_euclidean.py]: I ended up looping on the range so it would be easier to get the correspondent reference matrix, saved on a list, for each domain, although it may not be the best way to do it. I was thinking first on saving the reference matrices on dictionaries, indexed by the domain, but then I would need to set before the number of the test domain.

File: pyriemann/transfer/euclidean.py [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+
+        self.fit(X, y_enc)
+
+        _, _, domains = decode_domains(X, y_enc)
+
+        X_align = self._transform_sw(X, domains)
+
+        return X_align
+
+    def __sklearn_is_fitted__(self):
```
Comment (gcattan) [inline pyriemann/transfer/euclidean.py:256]: Your transformer always needs fitting. I don't think you can call transform without fit first. I suggest to remove this override.

Comment (brunaafl) [reply pyriemann/transfer/euclidean.py:256]: Yes, I agree! Supposing that you have data from different 'train' subjects (that will be used to train a classifier, for example) + calibration data from 'target' subject, and you want to use calibration data to compute reference for aligning the test data from 'target', the pipeline I was thinking was:

'Train' subjects --> `fit_transform` --> aligned 'Train'

'Calibration' --> `fit` calibration (compute reference)  or `fit_transform` if you want aligned calibration data

'Test' --> `transform` 'test' **after fitting calibration** --> aligned 'Test'

In this way, the last reference matrix computed/saved would be from calibration, and it would be used for transforming the test! I'm not sure if the way I implemented would be the best to do it, but that is what I was thinking

File: pyriemann/transfer/euclidean.py [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+
+    This transformation accept both raw signals and covariance matrices as
+    data inputs. For raw trials, you can determine the estimator using the
+    'estimtor' parameter.
+
+    .. note::
+       Using .fit() and then .transform() will give different results than
+       .fit_transform(). In fact, .fit_transform() should be applied on the
+       training dataset (target and source) and .transform() on the test
+       partition of the target dataset.
```
Comment (sylvchev) [inline pyriemann/transfer/euclidean.py:61]:
```suggestion
       If applied offline, that is to transform the training data, use .fit_transform().
       For pseudo-online or calibration setup, with training data from one domain
       (one subject) and unseen tested data, use .fit() and .transform(). You should
       ensure that there is only one domain in training data.
       
       Using .fit() and then .transform() will give different results than
       .fit_transform(). In fact, .fit_transform() should be applied on the
       training dataset (target and source) and .transform() on the test
       partition of the target dataset.
```

File: pyriemann/transfer/euclidean.py [hunk 18]
```diff
@@ -0,0 +1,258 @@
+import numpy as np
+
+from warnings import warn
+
+from scipy.linalg import sqrtm, inv
+from numpy import iscomplexobj, real, any, isfinite
+from sklearn.base import BaseEstimator, TransformerMixin
+from pyriemann.estimation import covariances
```
Comment (bruAristimunha) [inline pyriemann/transfer/euclidean.py:8]:
```suggestion
from ..estimation import covariances
```

**Review by qbarthelemy — CHANGES_REQUESTED** [review]
Thx @brunaafl for this contribution!

Due to their very high similarity, `EuclideanAlignment(dtype="covmat")` should be simply implemented using `TLCenter(target_domain="", metric="euclid")`, ie allowing an empty str for `target_domain` and modifying `transform` for this particular case (see https://github.com/pyRiemann/pyRiemann/blob/master/pyriemann/transfer/_estimators.py#L215).

**Comment by sylvchev** [comment]
Good call @qbarthelemy 
I discussed with @brunaafl , she will try to to rely on TLCenter for euclidean alignement as much as possible.

**Comment by qbarthelemy** [comment]
`transform` method of class `TLCenter` can be simply generalized into:
```
    def transform(self, X, y_enc):
        """Re-center matrices to Identity.

        If ``target_domain`` is not empty, recenter matrices to this
        specified target domain.
        Else, recenter matrices of each domain to Identity.

        Parameters
        ----------
        X : ndarray, shape (n_matrices, n_channels, n_channels)
            Set of SPD matrices.
        y_enc : ndarray, shape (n_matrices,)
            Extended labels for each matrix.
            Used only when ``target_domain`` is empty.

        Returns
        -------
        X : ndarray, shape (n_matrices, n_classes)
            Set of SPD matrices with mean in the Identity.
        """
        if self.target_domain != "":
            return self.recenter_[self.target_domain].transform(X)

        _, _, domains = decode_domains(X, y_enc)

        X_rct = np.zeros_like(X)
        for d in np.unique(domains):
            idx = domains == d
            X_rct[idx] = self.recenter_[d].transform(X[idx])
        return X_rct
```

**Comment by brunaafl** [comment]
Hi!

I made some changes to the TLCenter code, as @qbarthelemy  suggested.

I changed the parameter “target_domain” to be optional, allowing the same functionality as before to be maintained. 

```
    def transform(self, X, y_enc=None):
        """Re-center the data points in the target domain to Identity.

        Parameters
        ----------
        X : ndarray, shape (n_matrices, n_channels, n_channels)
            Set of SPD matrices.
        y_enc : None
            Not used, here for compatibility with sklearn API.

        Returns
        -------
        X : ndarray, shape (n_matrices, n_classes)
            Set of SPD matrices with mean in the Identity.
        """
        # Used during inference, apply recenter from specified target domain.

        # Therefore, if you specified before the target
        if self.target_domain != "":
            recenter = self.recenter_[self.target_domain].transform(X)

        # If you didn't specified, you must have fitted before on calibration
        else:
            keys = list(self.recenter_.keys())

            # Get last reference matrix (after fitting on the calibration)
            last = keys[-1]
            recenter = self.recenter_[last].transform(X)

        return recenter
```

However, if used as a transformation in a pipeline in a case where “target_domain” should be left empty, such as in an online cross-domain evaluation, you should fit the transformation to a calibration set of the target before transforming the test data. Doing that, the reference matrix computed using calibration will be the last one saved, and you will use it to recenter the test. 

With this structure, however, I’m not solving the problem of overriding the fit_transform, and applying `fit` and then `transform` will continue to be different from applying `fit_transform`, since this supposes that data passed in `transform` is from target subject. A way I imagined of solving this would be using the fact that y_enc==None in inference for the target, and not None otherwise.

Another thing I was imagining is if it was worth it to add a possibility of aligning “raw” trials as well, so it can be used in a more general way.

Please let me know what are your thoughts, and thank you all for all the comments until now!

**Comment by qbarthelemy** [comment]
Thank you @brunaafl  for this improvement of `TLCenter` class!
I just updated the doc and code. Don't forget to pull before continuing to code.

It's ok for me to use the last calibrated domain if `target_domain` is not defined.
That said, it seems that this does not cover the use case of He & Wu 2019...
On the other hand, to be sure that the user makes an explicit choice, `target_domain` must not be optional.

For recentering raw signals, I think it should rather be added to MOABB, because pyRiemann is more dedicated to SPD matrices.

**Review by qbarthelemy — APPROVED** [review]
Thx @brunaafl !
